docs/middleware.md

```diff
@@ -232,6 +232,252 @@ around explicitly, rather than mutating the middleware instance.
+```
+
+The `ASGI3Application` is meant to represent an ASGI application that follows the third version of the standard.
+Starlette itself is an ASGI 3 application.
+
+### Wrapping the request
+
+On Starlette, we have data structures that make it more convenient to organize your ASGI middleware.
+
+For example, we can create a `Request` object, and work with it.
+```python
+from starlette.requests import Request
+
+class ASGIMiddleware:
+    def __init__(self, app):
+        self.app = app
+
+    async def __call__(self, scope, receive, send):
+        if scope["type"] == "http":
+            request = Request(scope, receive, send)
+            # Do something here!
+        await self.app(scope, receive, send)
+```
+
+Also, it's possible to make use of `MutableHeaders` to change the headers:
+
+```python
+class ASGIMiddleware:
+    def __init__(self, app, headers):
+        self.app = app
+        self.headers = headers
+
+    async def __call__(self, scope, receive, send):
+        if scope["type"] == "http":
+            async def send_wrapper(message):
+                if message["type"] == "http.response.start":
+                    headers = MutableHeaders(scope=message)
+                    for key, value for self.headers:
+                        headers.append(key, value)
+                await send(message)
+        await self.app(scope, receive, send_wrapper)
+```
+
+### Per-request state
+
+The ASGI middleware class is meant to be stateless, as we don't want to spread the state from one request to another.
+The responder pattern is a way to deal with per-request state. The idea is to have an object that will hold that information.
+
+```python
+class ASGIMiddleware:
+    def __init__(self, app):
+        self.app = app
+
+    async def __call__(self, scope, receive, send):
+        if scope["type"] == "http":
+            responder = Responder(self.app)
+            return await responder(scope, receive, send)
+
+        await self.app(scope, receive, send)
+
+
+class Responder:
+    def __init__(self, app):
+        self.app = app
+        self.started = False
+
+    def __call__(self, scope, receive, send):
+        async def send_wrapper(message):
+            if message["type"] == "http.response.body" and not self.started:
+                self.started = True
+            await send(message)
+        await self.app(scope, receive, send_wrapper)
+```
+
+The `Responder` object, in this case, will save the information about when the response body has "started" to be sent.
+
+!!! info
+    You can see a more advanced implementation of this pattern on our [GZipMiddleware](), and also on the [msgpack-asgi](https://github.com/florimondmanca/msgpack-asgi/blob/6261b1e22b3689551f68038ee00adda5fbc04670/src/msgpack_asgi/_middleware.py).
+
+### Storing context in `scope`
+
+As we know by now, the `scope` holds the information about the application. To be precise, the `scope` holds
+the stateless information of the application. IS THIS CORRECT?
```

Comment:
Rather, the `scope` stores information about the current request. It's got information such as "headers", "path", "query", etc.

Reply:
"IS THIS CORRECT" was more about the wording. I'll reformulate the sentence.

Reply:
I think it would be "information about the connection" instead of about the application.

Maybe it could help to add an example in the explanation, as Florimond says, in the case of HTTP it includes information about one request, like the headers, path, etc. In the case of WebSockets, it also includes headers, paths, etc, but it's not a single request, instead, it's a WebSocket connection that can have multiple messages received and sent.